### PR TITLE
sharpe ratio

### PR DESCRIPTION
**Pull request by WojtekTB**

### CODE REVIEW

**Comment by WojtekTB** [comment]
FYI: untested

**Comment by WojtekTB** [comment]
According to [this](https://towardsdatascience.com/400x-time-faster-pandas-data-frame-iteration-16fb47871a0a) .apply() is very good, idk why but I am willing to believe it

**Comment by lucienqchen** [comment]
.apply is vectorized so it is going to be much faster than iterating through in base python, since it is written in C. However, you can do arithmetic operations directly on a series so there is no need to apply a lamda function i.e ((dly_returns - smt) **2).sum() would work just fine

File: sigpytch/metrics.py [hunk 1]
```diff
@@ -92,11 +91,35 @@ def volatility(price_data: pd.Series, window_width: int, window_index: int = 0)
...
+    """
+    returns_asset = return_daily(price_data_asset, window_width, window_index)
+    returns_benchmark = return_daily(price_data_benchmark, window_width, window_index)
+    T = window_width
+    D_t = pd.Series()
+    for return_asset, return_benchmark in returns_asset, returns_benchmark:
+        D_t.add(return_asset - return_benchmark)
+    D_avr = D_t.mean()
+
+    D_sigma = (D_t.apply(lambda D: (D-D_avr)**2).sum()/(T-1))**(1/2)
```
Comment (pranaysjha) [inline sigpytch/metrics.py:123]: can just use .std() here lol

File: sigpytch/metrics.py [hunk 2]
```diff
@@ -92,11 +91,35 @@ def volatility(price_data: pd.Series, window_width: int, window_index: int = 0)
...
+        raise Exception("Invalid window width and/or index. Window outside price data series...")
+    
     av_returns = return_average(price_data, window_width, window_index)
     dly_returns = return_daily(price_data, window_width, window_index)
-    T = len(dly_returns)
-    sigma = 0
-    for daily_return in dly_returns:
-        sigma += ((daily_return - av_returns)**2)/T
+    T = window_width
+    sigma = dly_returns.apply(lambda daily_return: ((daily_return - av_returns)**2)/T).sum()
```
Comment (pranaysjha) [inline sigpytch/metrics.py]: can use std() to compute sigma

File: sigpytch/metrics.py (hunk 1 above)
Comment (pranaysjha) [inline sigpytch/metrics.py]: can use volatility func instead here?

File: sigpytch/metrics.py [hunk 3]
```diff
@@ -92,11 +91,35 @@ def volatility(price_data: pd.Series, window_width: int, window_index: int = 0)
...
+    """
+    Computes sharpe_ratio over a period between asset and benchamrk asset. O(window_width)
+    
+    Parameters:
+        price_data_asset (pd.Series): Daily price data of asset
+        price_data_benchmark (pd.Series): Daily price data of benchmark asset
+        window_index (int): Start index of the window
+        window_width (int): Width of the window
+    Returns:
+        flaot: Sharpe ratio between asset and the benchmark asset
```
Comment (pranaysjha) [inline sigpytch/metrics.py]: typo

**Pull request closed without merging**